core/fcache.c

```diff
@@ -3898,14 +3909,50 @@ chain_fragments_for_flush(dcontext_t *dcontext, fcache_unit_t *unit, fragment_t
     } else {
         ASSERT(list != NULL);
         ASSERT(prev_f != NULL);
-        prev_f->next_vmarea = NULL;
     }
 
     if (tail != NULL)
         *tail = prev_f;
     return list;
 }
 
+bool
+fcache_flush_fragment(dcontext_t *dcontext, app_pc tag)
+{
+    bool did_flush = false;
+    fragment_t *list_head = NULL;
+    DEBUG_DECLARE(bool flushed;)
+
+    /* we pass false to flush_fragments_in_region_start() below for owning the initexit
+     * lock */
+    ASSERT_DO_NOT_OWN_MUTEX(true, &thread_initexit_lock);
+
+    /* Stage 1 of flushing: do synch. */
+    DEBUG_DECLARE(flushed =)
+    flush_fragments_synch_unlink_priv(dcontext, EMPTY_REGION_BASE, EMPTY_REGION_SIZE,
+                                      false /*don't have thread_initexit_lock*/,
+                                      false /*not invalidating exec areas*/,
+                                      false /*don't force synchall*/
+                                      _IF_DGCDIAG(NULL));
+    ASSERT(flushed);
+
+    fragment_t *f = fragment_lookup(dcontext, tag);
+    if (f != NULL && TEST(FRAG_SHARED, f->flags)) {
+        chain_single_fragment(dcontext, f, &list_head);
+        did_flush = true;
+    }
+
+    /* Stage 2 of flushing do the rest of the unlink and move the fragments to pending del
+     * list */
+    flush_fragments_unlink_shared(dcontext, EMPTY_REGION_BASE, EMPTY_REGION_SIZE,
```

Comment:
Add a check to not call this if list_head is NULL?  Will the vmareas code all handle the empty region?  Probably -- but I remember some of the recent issues about zero-length asserts and whatnot in some vmareas code.

Reply:
Hmm, I cannot just skip stage 2 as there are asserts in place to ensure it is performed.

Reply:
OK, it is just a little odd that flush_fragments_unlink_shared takes either a non-NULL list or a region, and here you want to pas a list and *not* a region, but you end up passing a region (empty) when list_head is NULL.

Reply:
Yes, I agree; it does seem weird. There is an existing comment that addresses this slightly: 

https://github.com/DynamoRIO/dynamorio/blob/cb1a6e028367943a76948e9ab8e18948a3e35771/core/fragment.c#L6542

Reply:
Maybe just add a comment here.